src/widgets/list.rs

```diff
@@ -1065,4 +1073,157 @@ mod tests {
+        .collect::<Vec<ListItem>>();
+        let list = List::new(items);
+        let buffer = render_widget(list, 7, 5);
+        let expected =
+            Buffer::with_lines(vec!["Odd    ", " Even  ", "  Width", "       ", "       "]);
+        assert_buffer_eq!(buffer, expected);
+    }
+
+    #[test]
+    fn test_render_list_alignment_even_line_even_area() {
+        let items = [
+            Line::from("Odd").alignment(Alignment::Left),
+            Line::from("Even").alignment(Alignment::Center),
+            Line::from("Width").alignment(Alignment::Right),
+        ]
+        .into_iter()
+        .map(ListItem::new)
+        .collect::<Vec<ListItem>>();
+        let list = List::new(items);
+        let buffer = render_widget(list, 6, 4);
+        let expected = Buffer::with_lines(vec!["Odd   ", " Even ", " Width", "      "]);
+        assert_buffer_eq!(buffer, expected);
+    }
+
+    #[test]
+    fn test_render_list_alignment_odd_line_even_area() {
+        let items = [
+            Line::from("Odd").alignment(Alignment::Left),
+            Line::from("Even").alignment(Alignment::Center),
+            Line::from("Width").alignment(Alignment::Right),
+        ]
+        .into_iter()
+        .map(ListItem::new)
+        .collect::<Vec<ListItem>>();
+        let list = List::new(items);
+        let buffer = render_widget(list, 8, 4);
+        let expected = Buffer::with_lines(vec!["Odd     ", "  Even  ", "   Width", "        "]);
+        assert_buffer_eq!(buffer, expected);
+    }
+
+    #[test]
+    fn test_render_list_alignment_even_line_odd_area() {
+        let items = [
+            Line::from("Odd").alignment(Alignment::Left),
+            Line::from("Even").alignment(Alignment::Center),
+            Line::from("Width").alignment(Alignment::Right),
+        ]
+        .into_iter()
+        .map(ListItem::new)
+        .collect::<Vec<ListItem>>();
+        let list = List::new(items);
+        let buffer = render_widget(list, 6, 5);
+        let expected = Buffer::with_lines(vec!["Odd   ", " Even ", " Width", "     ", "     "]);
+        assert_buffer_eq!(buffer, expected);
+    }
+
+    #[test]
+    fn test_render_list_alignment_zero_line_width() {
+        let items = [Line::from("This line has zero width").alignment(Alignment::Center)]
+            .into_iter()
+            .map(ListItem::new)
+            .collect::<Vec<ListItem>>();
+        let list = List::new(items);
+        let buffer = render_widget(list, 0, 5);
+        let expected = Buffer::with_lines(vec!["", "", "", "", ""]);
+        assert_buffer_eq!(buffer, expected);
+    }
+
+    #[test]
+    fn test_render_list_alignment_zero_area_width() {
+        let items = [Line::from("Text").alignment(Alignment::Left)]
+            .into_iter()
+            .map(ListItem::new)
+            .collect::<Vec<ListItem>>();
+        let list = List::new(items);
+        let buffer = render_widget(list, 4, 0);
+        let expected = Buffer::with_lines::<&str>(vec![]);
```

Comment:
Yeah, this is a weird one. assert_buffer_eq! has problems with buffers with zero height. It's probably best to just handle it by avoiding the issue:

```suggestion
        // assert_buffer_eq! doesn't handle zero height buffers so we call this test manually
        // rather than using render_widget
        let mut buffer = Buffer::empty(Rect::new(0, 0, 4, 1));
        Widget::render(list, Rect::new(0, 0, 4, 0), &mut buffer);
        let expected = Buffer::with_lines(vec!["    "]);
```

Reply:
f247e48ed6bd06fc944514344652dffe1e679903